modules/nat64/dataplane.c

```diff
@@ -0,0 +1,413 @@
+	new_ipv4_header->total_length =
+		rte_cpu_to_be_16(payload_length + sizeof(struct rte_ipv4_hdr));
+
+	new_ipv4_header->packet_id = 0;	      // TODO: generate id
+	new_ipv4_header->fragment_offset = 0; // TODO: handle fragmentation
+	new_ipv4_header->time_to_live =
+		ipv6Header->hop_limits; // TODO: decrement ttl?
+	new_ipv4_header->next_proto_id = ipv6Header->proto;
+	new_ipv4_header->hdr_checksum = 0;
+
+	memcpy(&new_ipv4_header->src_addr, new_src_addr->ip4, sizeof(uint32_t));
+	// memcpy(&new_ipv4_header->dst_addr, new_dst_addr->ip4, sizeof(struct
+	// in_addr));
+
+	// handle ICMP, TCP, UDP
+	if (ipv6Header->proto == IPPROTO_ICMPV6) {
+		new_ipv4_header->next_proto_id = IPPROTO_ICMP;
+
+		struct icmp6_hdr *icmpHeader = rte_pktmbuf_mtod_offset(
+			mbuf,
+			struct icmp6_hdr *,
+			packet->transport_header.offset
+		);
+		int result = icmp_v6_to_v4(icmpHeader);
+		if (result) {
+			RTE_LOG(ERR, NAT64, "icmp_v6_to_v4 failed\n");
+			return result;
+		}
+	}
+	// copy l2 header
+	rte_memcpy(
+		rte_pktmbuf_mtod_offset(mbuf, char *, delta),
+		rte_pktmbuf_mtod(mbuf, char *),
+		packet->network_header.offset
+	);
+
+	// reduce packet
+	if (rte_pktmbuf_adj(mbuf, delta) == NULL) {
+		RTE_LOG(ERR, NAT64, "adjust mbuf failed. Delta: %d\n", delta);
+		return -1;
+	}
+
+	// adjust new transport header offset
+	packet->transport_header.offset =
+		packet->network_header.offset + sizeof(struct rte_ipv4_hdr);
+
+	// set ipv4 header type
+	uint16_t *next_header_type = rte_pktmbuf_mtod_offset(
+		mbuf, uint16_t *, packet->network_header.offset - 2
+	);
+	*next_header_type = rte_cpu_to_be_16(RTE_ETHER_TYPE_IPV4);
+
+	// Обновляем длину пакета в структуре mbuf
+	mbuf->pkt_len -= 20;
+	mbuf->data_len -= 20;
+
+	return 0;
+}
+
+static int
+nat64_handle_v4(struct ip4to6 **hash, struct packet *packet) {
+
+	struct rte_mbuf *mbuf = packet_to_mbuf(packet);
+
+	struct rte_ipv4_hdr *ipv4Header = rte_pktmbuf_mtod_offset(
+		mbuf, struct rte_ipv4_hdr *, packet->network_header.offset
+	);
+
+	uint32_t new_packet_len =
+		mbuf->pkt_len + sizeof(struct in6_addr) - sizeof(uint32_t);
+
+	uint16_t delta =
+		packet->transport_header.offset -
+		packet->network_header.offset +
+		(sizeof(struct rte_ipv6_hdr) - sizeof(struct rte_ipv4_hdr));
+
+	rte_pktmbuf_prepend(
+		mbuf, sizeof(struct rte_ipv6_hdr) - sizeof(struct rte_ipv4_hdr)
+	);
+	struct rte_ipv6_hdr *new_ipv6_header = rte_pktmbuf_mtod_offset(
+		mbuf, struct rte_ipv6_hdr *, packet->network_header.offset
+	);
+
+	rte_memcpy(
+		rte_pktmbuf_mtod(mbuf, char *),
+		rte_pktmbuf_mtod_offset(mbuf, char *, delta),
+		packet->network_header.offset
+	);
+
+	new_ipv6_header->vtc_flow = ipv4Header->type_of_service;
+	new_ipv6_header->payload_len = rte_cpu_to_be_16(new_packet_len);
+	new_ipv6_header->hop_limits =
+		ipv4Header->time_to_live; // TODO: decrement?
+	new_ipv6_header->proto = ipv4Header->next_proto_id;
+
+	uint32_t addr4 = rte_be_to_cpu_32(ipv4Header->dst_addr);
+	struct ip4to6 *entry = find_ip4to6(hash, &addr4);
```

Comment:
Move the check immediately after retrieving the IPv4 header to:
- Prevent unnecessary processing
- Avoid reading invalid data post-mbuf modifications